### ISSUE TITLE
[css-view-transitions-2] Ignore offscreen elements from participating in transitions

### ISSUE THREAD

**Issue by khushalsagar** [w3c/csswg-drafts#8282]
Currently if an element has a non-none computed value for `view-transition-name`, it participates in the transition irrespective of whether it is in the visible viewport. This means the element will be rendered, which has significant computational and memory overhead, even if it is never seen by the user. If the developer wants to avoid this overhead, they have to keep track of the visibility of each element and only add `view-transition-name` to the onscreen ones.

The proposal to make this case easier is as follows:

- Add a new CSS property view-transition-offscreen which customizes the behavior for named elements based on their position in the snapshot viewport.
- The property has 2 values: auto and absent. auto indicates that the UA should render the element irrespective of its viewport position (as-if its onscreen). Would be nice to allow flexibility to the UA to optimize out such elements in case the transition is on memory constrained devices. That's why "should" instead of "must".
- absent indicates that if the element's ink overflow rectangle does not intersect the snapshot viewport, the element does not participate in the transition (as-if its view-transition-name's computed value was none).
- The computation for the element's viewport position for the decision above can be done [here](https://drafts.csswg.org/css-view-transitions-1/#perform-pending-transition-operations-algorithm:~:text=associated%20document.-,For%20each%20element%20of%20every%20Element%20and%20pseudo%2Delement%20connected%20to%20document%2C%20in%20paint%20order%3A,-The%20link%20for) for old elements. And [here](https://drafts.csswg.org/css-view-transitions-1/#perform-pending-transition-operations-algorithm:~:text=for%20transition.-,Update%20pseudo%2Delement%20styles%20for%20transition.,-Note%3A%20The) for new elements. The subtlety is that for new elements we do it before resolving the ready promise so script can observe the final pseudo DOM structure.

See prior discussion on this [here](https://github.com/WICG/view-transitions/issues/84).

**Comment by jakearchibald** [comment]
There are memory and computational benefits to this feature, but I originally proposed this as more of a visual / developer experience feature.

If two pages have a common heading, you may want that to be static in a transition, so you give it a `page-transition-name`.

However, if one of the pages is scrolled 8000px, the transition between the two will be bad, as the header will fly in from 8000px away.

With this feature, developers will be able to create a special incoming/outgoing animation for the header in this case (using the `:only-child` selector).

**Comment by jakearchibald** [comment]
> * The property has 2 values: auto and absent. auto indicates that the UA should render the element irrespective of its viewport position (as-if its onscreen). Would be nice to allow flexibility to the UA to optimize out such elements in case the transition is on memory constrained devices. That's why "should" instead of "must".

Since it significantly changes the animation, I'm not sure we can, or should, do this automatically. Since we won't be doing it in the first release, and it will only happen on constrained devices, I think it'll lead to things appearing broken when this 'auto' behaviour kicks in. I'd rather say that transitions can be skipped if the device is constrained, since that's a more reliable fallback that developers will already be catering for.

That said, I support the default value being `auto`, since elements are sometimes ignored due to `content-visibility` https://github.com/w3c/csswg-drafts/issues/7874

**Comment by vmpstr** [comment]
It feels like this is one of these options that you should give to startViewTransition to optimize things because you know that you've marked way too much with the view-transition-name. Although I do see the appeal of the new css property since you can use that in MPA cases too, to me it doesn't really feel like the right abstraction

**Comment by jakearchibald** [comment]
I like it as a CSS property since the instruction can be per transition item. Eg, if you're reordering a list, you still want things to transition from outside the viewport. Whereas you might not want that behaviour for a header. Both might happen in the same transition.

**Comment by khushalsagar** [comment]
A third value which would be useful for a case like https://deploy-preview-32--infrequently.netlify.app/, the page has a massive element which will be captured in entirety while the animation doesn't need that.

Syntax ideas, `view-transition-offscreen: clip` allows the UA to snapshot an intersection of the element's ink overflow rectangle with the visible viewport or [snapshot root](https://drafts.csswg.org/css-view-transitions-1/#snapshot-root-concept). Potential add-on is `view-transition-offscreen: clip inset(10px)`, second argument allows specifying an explicit skirt by which the snapshot should be expanded in either direction.

**Comment by jakearchibald** [comment]
Isn't the clipping just supposed to happen automatically? https://drafts.csswg.org/css-view-transitions-1/#compute-the-interest-rectangle-algorithm

**Comment by khushalsagar** [comment]
> Isn't the clipping just supposed to happen automatically? https://drafts.csswg.org/css-view-transitions-1/#compute-the-interest-rectangle-algorithm

That automatic clipping is very conservative, only if we must because of constraints like max texture size. Technically an implementation doesn't need to be constrained by it (you could create a tiled image), but the option gives UA flexibility.

This property would be an explicit hint from the developer that only the onscreen content (or a skirt around it) of this DOM element will be animated during the transition. The UA can use this knowledge to aggressively optimize for memory by painting and snapshotting a subset of the DOM element.

**Comment by bramus** [comment]
On a hackathon I coached at this was a dealbreaker for some, as they saw performance get tanked on non-highend devices. They were animating 1 element out of a list of 50, of which only 7 of them where visible in the viewport. Having a way to easily exclude these offscreen elements would surely be beneficial here.

**Comment by jakearchibald** [comment]
Are they happy with some items being :only-child, and animating as such, even though they existed in both states.

I wonder if we need some other feature in that case, where the group still animates from old to new, but there's only a new in the pair, or the group is dropped if the final position of the group is still out of view.

**Comment by khushalsagar** [comment]
> I wonder if we need some other feature in that case, where the group still animates from old to new, but there's only a new in the pair, or the group is dropped if the final position of the group is still out of view.

@jakearchibald we have a couple of other issues with the characteristics you mentioned:

- https://github.com/w3c/csswg-drafts/issues/9354 does the "group is dropped if the final position of the group is still out of view". This is helpful for cases where you want an element to either morph (if its onscreen both sides) or undo capturing it on the old side.
- https://github.com/w3c/csswg-drafts/issues/9406 does the "the group still animates from old to new, but there's only a new in the pair". The browser can optimize quite a lot with this. We are forced to capture images for offscreen elements in the old DOM elements (because they will be destroyed) but there's no such requirement for the new DOM. If the pseudo displaying that image is not onscreen, we don't need to render it.

We haven't dug into the exact API shape but given how related these 3 features are, I feel like we should tackle them together. Like a new CSS property to specify one of these modes?

**Comment by noamr** [comment]
I like the direction this conversation was going. The main thing that justifies a new attribute for this is the idea that this could be a UX choice rather than a mere optimization (e.g. preventing a header from jumping).

I think the semantics here should be similar to content-visibility and intersection observers, however because this is observable and not just an optimization, the definitions need to be exact and customizable via a margin (same as `rootMargin` in `IntersectionObserver`).

Another thing I think we should do is make this new attribute inherited, this way the author can decide that a container makes its entire set of descendants behave in a certain way (and this can be overridden further down the tree).

Perhaps this can be `view-transition-visibility` or `view-transition-overflow` with:
- `visible`: current behavior
- `auto`: similar to `content-visibility: auto`. element acts as if it doesn't have a `view-transition-name` if it's not intersecting with the viewport.
- `hidden`: element ignores its `view-transition-name`. As this attribute is inherited, this can be used to hide an entire tree from the view transitioning capture algorithm
- `clip`: the element participates in the transition, but only contents of the element that intersect with the viewport are captured..

**Comment by khushalsagar** [comment]
+1 to making the property inherited. I'm assuming the initial value will be `visible`.

For the `auto`/`clip` case, is there any use-case to let developers expand the snapshot viewport by some margin. So any elements/area within that is considered visible. This can also be a future extension with the current capability limited to clipping at snapshot viewport boundary.

**Comment by khushalsagar** [comment]
> I wouldn't do anything regarding the ink overflow. It shouldn't be web observable... so I think the intersection should work according to the IntersectionObserver rules and disregard ink-overflow.

Hmmm, I think there will be cases where disregarding the ink overflow would end up with glitchy behaviour. Like a widget whose shadow is in the viewport. Should we really ignore it? This has come up for IO in the past too: https://github.com/w3c/csswg-drafts/issues/8649. @szager-chromium on that.

I was curious how content-visibility handles this. Looks like it relies on `overflow-clip-margin` (which is developer provides) and uses that [here](https://drafts.csswg.org/css-contain/#close-to-the-viewport:~:text=The%20element%20is%20close%20to%20the%20viewport%3A%20In%20this%20state%2C%20the%20element%20is%20considered%20%22on%2Dscreen%22%3A%20its%20paint%20containment%20box%27s%20overflow%20clip%20edge%20intersects%20with%20the%20viewport%2C%20or%20a%20user%2Dagent%20defined%20margin%20around%20the%20viewport.). Even if the overflow clip edge is further than the actual ink or scroll overflow, we rely on the specified edge.

With VT we intentionally decided not to have paint containment, so that won't work for us.

> Not sure if clip should be something a different attribute. When the whole content is clipped-out, do you want to treat it as an empty image or not have it at all?

Oh I just read what you said carefully and I tend to agree. We should have separate properties which decide whether the element participates in the transition. If it is participating, then a separate property has knobs for deciding how its captured. So omitting `clip` sounds good.

In that regard, maybe the use-case in https://github.com/w3c/csswg-drafts/issues/9354 should eventually be handled by `view-transition-visibility`.

**Comment by jakearchibald** [comment]
> I think there will be cases where disregarding the ink overflow would end up with glitchy behaviour.

Also agree. I think ink-overflow intersection with the viewport sounds as visible in terms of a view transition, else you risk 'seeing double'.

**Comment by vmpstr** [comment]
> I was curious how content-visibility handles this.

Yeah, content-visibility uses paint containment for this reason, so that we don't need to render the subtree to figure out the extent of the overflow, all of the needed information is on the box itself (border box + overflow-clip-margin)

**Comment by khushalsagar** [comment]
One of the open questions for this feature is defining when an element is considered visible. There's 2 aspects to it:

- Intersection with the viewport (snapshot root from VT perspective). Ideally we'd use the element's ink overflow rect for it. IntersectionObserver already uses ink overflow rect to detect occlusion [here](https://w3c.github.io/IntersectionObserver/v2/#:~:text=Implementations%20should%20use%20the%20ink%20overflow%20rectangle%20of%20page%20content%20when%20determining%20whether%20a%20target%20is%20occluded). And #8649 is in progress to make the ink overflow bounds interoperable.

- Occlusion by other elements on the page. I haven't seen a use-case where a named element was occluded so I don't know if we need to consider it. Occlusion calculations are also more expensive so better if we can avoid it.

**Comment by jakearchibald** [comment]
I'm pretty certain we don't need to care about occlusion. Just viewport intersection.

**Comment by noamr** [comment]
> One of the open questions for this feature is defining when an element is considered visible. There's 2 aspects to it:
> 
> 
> 
> - Intersection with the viewport (snapshot root from VT perspective). Ideally we'd use the element's ink overflow rect for it. IntersectionObserver already uses ink overflow rect to detect occlusion [here](https://w3c.github.io/IntersectionObserver/v2/#:~:text=Implementations%20should%20use%20the%20ink%20overflow%20rectangle%20of%20page%20content%20when%20determining%20whether%20a%20target%20is%20occluded). And #8649 is in progress to make the ink overflow bounds interoperable.
> 
> 
> 
> - Occlusion by other elements on the page. I haven't seen a use-case where a named element was occluded so I don't know if we need to consider it. Occlusion calculations are also more expensive so better if we can avoid it.

In either case I think we should be consistent with IntersectionObserver. For viewport intersection IntersectionObserver doesn't take the ink overflow into account, so this shouldn't either. If IO exposed ink overflow in some way, eg for occlusion, we could consider doing the same.  

Note that with a big enough root margin, being accurate about the ink overflow becomes less important.

**Comment by jakearchibald** [comment]
hmm, I think the developer intent of this feature trumps consistency with intersection observer here. We can avoid using the word 'intersection' if that's where the problem is.

**Comment by noamr** [comment]
> hmm, I think the developer intent of this feature trumps consistency with intersection observer here. 

How is the developer intent here different from the developer intent in `IntersectionObserver`?
I think consistency is important here, having several APIs that rely on viewport-intersection and work in a slightly different manner would create confusing UX bugs and confusion.

**Comment by noamr** [comment]
I think the underlying question here is how we would expect elements to behave that are right outside the edge of the viewport.
If this was just about performance, we could make some educated guess and tweak it. But this about excluding elements that are outside the viewport from participating. 

So let's say you have an element that has a `top: 100vh` or some such - deliberately right outside the viewport. Should a 1px blur make it suddenly participate in the transition? Feels to me that this should not be something that relies on implementation-specific things and the author should be able to curate this with properties rather than rely on this unspeced behavior.

**Comment by jakearchibald** [comment]
> How is the developer intent here different from the developer intent in `IntersectionObserver`?

`IntersectionObserver` is more about reacting to an element when it's in the viewport, that's why it has options for being completely in the viewport, and everything in between. It was originally created with lazy-loading in mind.

With view transitions, we're in a very visual space. So "not there" becomes more of a visual question than a layout question. If we count "partially visible" as "not there", then you'll end up with transitions where the old thing and the new thing are both visible (to some degree) in the transition, but they won't transition as part of the same group, so the user will see double. I don't think that's a desirable outcome of this feature.

Also, parts of intersection observer _do_ consider ink overflow.

**Comment by noamr** [comment]
> Also, parts of intersection observer _do_ consider ink overflow.

There's an open proposal about occlusion, is that what you mean? Otherwise can you be specific?

**Comment by jakearchibald** [comment]
The occlusion stuff is shipped in Chrome, no?

**Comment by noamr** [comment]
> With view transitions, we're in a very visual space. So "not there" becomes more of a visual question than a layout question. If we count "partially visible" as "not there", then you'll end up with transitions where the old thing and the new thing are both visible (to some degree) in the transition, but they won't transition as part of the same group, so the user will see double. I don't think that's a desirable outcome of this feature.

I can totally see this point. The other side of this is that due to the implementation-specific nature of ink overflow, the outcome of using it here would be that in some cases you'd have elements that are visually 100% outside the viewport (on the edge) participate in the transition even if you gave them `view-transition-visibility: auto` - let's say they have a few pixels of blur that are reserved but not actually rendered, and you'd have no recourse to fixing it (except for maybe using `IntersectionObserver` yourself). Having a property for this (some viewport margin or element clip margin is enough) gives this control to the developer rather than to implementation details that can change at any time.

**Comment by noamr** [comment]
One thing we can do rather than rely on ink overflow specifically but would have the same effect, is to say that elements outside the viewport have an implementation-defined margin around them, which can be overridden with a CSS property.
This way if the author wants a very specific control of how this works they can do that, and the default is known to be implementation-defined.

**Comment by toddpadwick** [comment]
This would be really useful.  running into quite a few issues where this would drastically improve transitions. upvote from me 👍

**Comment by noamr** [comment]
> This would be really useful. running into quite a few issues where this would drastically improve transitions. upvote from me 👍

Would this be more about a performance optimization for you, or for design (avoiding things flying in/out of the screen)?

**Comment by toddpadwick** [comment]
Hey @noamr mainly design but performance is an issue as well, since if you're applying transform transitions to a whole page, and if the page is really long that can have cause huge issues on performance, and unexpected rendering consequences.
In terms of design:
1. Just as you've flagged, without having to do hacky functions to add view transition names based on intersection observer (which I think just takes away to beauty of how wonderfully simple the view transition API has the potential to be), things fly in and out unintentionally.
2. Lets say you wish to slide in the old view followed by the new view, a bit like a slideshow, you can't (unless I am missing something), because the whole page content is included, not just the clipped viewport content. See mock up below:
![Screenshot 2024-02-19 at 16 33 33](https://github.com/w3c/csswg-drafts/assets/36191462/edb1a216-d439-46af-859c-3215f6f3c02b)

**Comment by toddpadwick** [comment]
Really? I may be confused still but seems like its still relevant - since I do not want anything outside the viewport from participating in the transition - i want the transition-view image to just be the bounding box of the viewport. If I change the height of them, it compresses the whole page transition view to be 100vh, which is not right. 

eg this is the code I hope would achieve desired result:
```
::view-transition-old(root) {
    animation:slide-view-out 1s;
}

::view-transition-new(root) {
   animation:slide-view-in 1s;
}

@keyframes slide-view-out {
    0% {
      transform:translateY(0);
    }
    100% {
      transform:translateY(-100vh);
    }
  }

  @keyframes slide-view-in {
    0% {
      transform:translateY(100vh);
    }
    100% {
      transform:translateY(0);
    }
  }

```

**Comment by noamr** [comment]
> Really? I may be confused still but seems like its still relevant - since I do not want anything outside the viewport from participating in the transition

"Participating in the transition" is about whether the elements themselves participate individually, rather than whether the content is clipped and the size of the pseudo-element adjusting.

> i want the transition-view image to just be the bounding box of the viewport. If I change the height of them, it compresses the whole page transition view to be 100vh, which is not right.

I think `object-fit: cover` would help here? (Need to try some of this, added a note...)

**Comment by toddpadwick** [comment]
ah I see what you mean. Just left the office but will give that object-fit cover a go in the morning :) 
thanks @noamr

**Comment by toddpadwick** [comment]
Just to let you know, I gave that object-fit cover a go @noamr - doesn't seem to work – still has the same issue where it shrinks the height of the whole page (so you see a flash of the distorted page for a split second. I'm going to revert to using Vue JS page transitions for the time being.

**Comment by bokand** [comment]
@toddpadwick - do you have an example URL of the "slideshow" transition not working (apologies if it's in the above URL but I can't find it...)? The root snapshot should be limited to what's in the viewport and I've used this approach in some toy examples (https://viewtransition.glitch.me/test/gesturedemo-vertical.html).

The only thing I could think of is on mobile, if your site is wider than the device screen but you have `initial-scale=1` in your viewport meta you're technically zoomed in so the "captured viewport" is larger than your visual viewport. But that's not a consideration on desktop if that's where you're seeing it. If you have a link I can take a closer look...

**Comment by khushalsagar** [comment]
@toddpadwick the spec explicitly clips the root snapshot to the viewport, see text [here](https://drafts.csswg.org/css-view-transitions-1/#capture-the-image-algorithm:~:text=Render%20the%20region%20of%20document%20(including%20its%20canvas%20background%20and%20any%20top%20layer%20content)%20that%20intersects%20the%20snapshot%20containing%20block). If you have a case where this is not happening for the root element, it's an implementation bug.

That said, the use-case you mentioned is still relevant for non-root elements and is being tracked in https://github.com/w3c/csswg-drafts/issues/9481. This issue is about whether an element participates in the transition and #9481 is about what portion of its content is captured in an image *if it participates in the transition*.

**Comment by noamr** [comment]
A potential name for this: `view-transition-cutoff: viewport | none`.

**Comment by jakearchibald** [comment]
`cutoff` feels like it could mean clipping, whereas this is about excluding.

I'm not saying any of these are better, but here goes nothing:

- `view-transition-requirements: viewport-intersection`
- `view-transition-exclude: outside-viewport`

ugh.

**Comment by jakearchibald** [comment]
Developers want this feature to exclude items that don't render within the viewport, and that includes considering ink overflow. I think it'd be hard to explain to them that they can't have the feature they want because that concept is too hard to spec 😄

**Comment by noamr** [comment]
> Developers want this feature to exclude items that don't render within the viewport, and that includes considering ink overflow. I think it'd be hard to explain to them that they can't have the feature they want because that concept is too hard to spec 😄

Looking at https://github.com/w3c/csswg-drafts/issues/8649#issuecomment-2010712895, there is a security issue with using the ink overflow as is, as things like a11y extensions can affect it (e.g. adding an outline/background to fonts that is not web-observable).

**Comment by jakearchibald** [comment]
Yeah, that's fair

**Comment by noamr** [comment]
> `cutoff` feels like it could mean clipping, whereas this is about excluding.
> 
> I'm not saying any of these are better, but here goes nothing:
> 
> * `view-transition-requirements: viewport-intersection`
> * `view-transition-exclude: outside-viewport`
> 
> ugh.

How about `view-transition-scissor: viewport | document` ? It's kind of like `glScissor`. Or `view-transition-extents: viewport`

**Comment by bramus** [comment]
> How about `view-transition-scissor: viewport | document` ?

Iterating on this:

- For the property name: `view-transition-snapshot-area`.
- For the values: `viewport | root`, with `root` being the VT root (which is `document` or, in the case of a scoped transition, the element on which the author called `.startViewTransition()`)

    `root` could be confusing as one might think it’s `:root`. I have considered `element` but that doesn’t map well to `document`. Another alternative for this could be `scope`.

    Also: what should be the default value?

**Comment by noamr** [comment]
> > How about `view-transition-scissor: viewport | document` ?
> 
> Iterating on this:
> 
> * For the property name: `view-transition-snapshot-area`.
> * For the values: `viewport | root`, with `root` being the VT root (which is `document` or, in the case of a scoped transition, the element on which the author called `.startViewTransition()`)
>   `root` could be confusing as one might think it’s `:root`. I have considered `element` but that doesn’t map well to `document`. Another alternative for this could be `scope`.
>   Also: what should be the default value?

I think today it's `root`? Per spec we don't capture elements that go beyond the document rect.

**Comment by jakearchibald** [comment]
I feel `scissor` has the same issue as `cutoff` - it feels like the exclusion could be partial. Whereas this feature is about the element being included or excluded from capture.

**Comment by noamr** [comment]
`view-transition-inclusion-area: viewport` ? A bit verbose :) Anyway, let's deal with the bikeshedding once we figure out the main issue (extents/ink-overflow).

**Comment by jakearchibald** [comment]
That seems ok. Although I'm not sure how you'd add the "but consider ink overflow" option in future.

**Comment by noamr** [comment]
> That seems ok. Although I'm not sure how you'd add the "but consider ink overflow" option in future. (I realise I'm not being a lot of help here)

I think in the future we can add a `view-transition-inclusion-edge` property that defaults to `border-box` and can be `ink-box` (or some such)

**Comment by noamr** [comment]
OTOH, I'm not sure if we want to extend this in the future to enable exclusive cutoff rather than inclusive, or even clipping (only the rect that's part of the viewport is snapshotted).

Perhaps this can be (extended in the future to) 4 properties, with a shorthand (replace the word `cutoff` with any other word that we choose):
```
view-transition-cutoff-area: viewport | root;
view-transition-cutoff-rule: inclusive | exclusive | clip; /* this is somewhat similar to canvas/svg fill-rule */
view-transition-cutoff-margin: length-percentage;
view-transition-cutoff-extents: border-box | paint-box | ...;

view-transition-cutoff: inclusive viewport; /* shorthand to include any element that intersects with the viewport */
view-transition-cutoff: inclusive root; /* this is the default */
view-transition-cutoff: clip viewport; /* this is the default for the root element*/
```

**Comment by khushalsagar** [comment]
For deciding whether an element's box is visible or not, we should probably use the [snapshot containing block](https://drafts.csswg.org/css-view-transitions-1/#snapshot-containing-block-concept). Since any part of that block can become visible during the transition (URL bar hidden vs shown) so we want to capture an element if its in that block.

Re: which reference rect to use, I agree using ink overflow would be ideal but +1 to start off with the simpler approach of using the border-box given the complexity with ink overflow. There's probably more cases where authors would be ok with using the border box (ignore the element if only a small shadow is showing).

Anchor positioning is running into this too (see https://github.com/w3c/IntersectionObserver/issues/522). Hopefully having use-cases come up from both these features will give us the motivation to figure out exposing ink overflow too. @fantasai @tabatkins fyi.

**Comment by noamr** [comment]
> For deciding whether an element's box is visible or not, we should probably use the [snapshot containing block](https://drafts.csswg.org/css-view-transitions-1/#snapshot-containing-block-concept). Since any part of that block can become visible during the transition (URL bar hidden vs shown) so we want to capture an element if its in that block.

Absolutely

**Comment by nt1m** [comment]
I would prefer if the default value gave the UA flexibility to apply whatever behavior it judges more efficient. Ideally developers wouldn't need to touch this.

**Comment by noamr** [comment]
> I would prefer if the default value gave the UA flexibility to apply whatever behavior it judges more efficient. Ideally developers wouldn't need to touch this.

That already exists, but this feature is more about allowing the author to use this for art-direction, like decide that if an element is offscreen in the old state it should/shouldn't fly in for the new state.

**Comment by khushalsagar** [comment]
> I would prefer if the default value gave the UA flexibility to apply whatever behavior it judges more efficient.

The current spec technically doesn't. The browser has to render offscreen content because the animation might bring it onscreen. For example, you delete an item from the list and the next item slides onscreen into place while its content (item #) animates.

But we can allow the browser to ignore elements which are away from the viewport given a UA defined margin. That's likely what authors want anyway, otherwise its only the last few frames of the animation when the element comes onscreen which looks weird.

So this property can be a tri-state of: visible, hidden and auto where auto uses UA heuristic to decide whether an element is visible or hidden.

**Comment by nt1m** [comment]
I can see a use for hidden, but not sure about visible, auto should be smart enough to do the right thing given the UA knows about the start and end states. Don't really have a super strong opinion here though

**Comment by nt1m** [comment]
> Capture if the element is onscreen. If it's offscreen, the UA can skip based on heuristics of how far offscreen it is.

I was thinking the UA could be smarter and detect if there would be any visible effect to omitting the animation based on the position of start+end state, but might not be trivial, which would make visible technically not useful. I'm fine either way.

**Comment by khushalsagar** [comment]
> I was thinking the UA could be smarter and detect if there would be any visible effect to omitting the animation based on the position of start+end state

That's the fundamental problem which requires a web API here. The browser can't know what the end state will be when its caching the old DOM, since it doesn't know what the new DOM is yet.

**Comment by khushalsagar** [comment]
At [this point](https://drafts.csswg.org/css-view-transitions-1/#capture-old-state-algorithm:~:text=Set%20capture%E2%80%99s%20old%20image%20to%20the%20result%20of%20capturing%20the%20image%20of%20element.) in the algorithm when we cache the old state, we just know that the named element in the old DOM is offscreen. We don't yet know if its corresponding named element in the new DOM will be onscreen. And when we get to new state steps, the old DOM has already been replaced by the new DOM.

I could imagine an optimized implementation internally preserves a cheap copy of the old DOM; such that generating an image of the old DOM elements can be deferred until the browser knows what the new DOM looks like. And skip rendering/caching the image if the new state is offscreen without any web observable effects. Hmmm, can webkit do this? :)

**Comment by jakearchibald** [comment]
> this feature is more about allowing the author to use this for art-direction

As an author, I want to +💯 this.

I don't think heuristics work here. By the time the browser knows the full animation for a group, it's already captured the old view. This feature is about being able to exclude the old view. Also, even if the animation won't intersect with the viewport at the start of the animation, there's no guarantee that will remain true - the DOM can change.

As an author, sometimes I'm happy with an element animating from outside the viewport to outside the viewport. Sometimes I'm happy for it to animate from its original position, outside the viewport, to its new position inside the viewport, or vice-versa. But sometimes I'm not. It depends on the thing being animated.

If that design choice is taken away from me, the feature is useless. This feature is about design choice.

**Comment by khushalsagar** [comment]
> I think the current behaviour is the right default. Same as CSS animations - we don't cancel out-of-viewport animations.

The fact that out-of-viewport CSS animations don't add rendering cost is a reason to consider a different default here. I especially feel that after seeing examples like https://github.com/w3c/csswg-drafts/issues/8320#issuecomment-2023077559. For an infinite scroller type of case, authors will add a name to all list items without realizing its causing the browser to render offscreen content. If they do need it to render (for cases you mentioned) and the browser doesn't by default, they'd notice a visually incorrect animation and opt-in to it.

But ok to take either approach at the WG meeting.

**Comment by jakearchibald** [comment]
You're always going to have the initial capture cost, because you need to do that before the 'end' position is given (but again, the end position can change throughout the animation).

I think that example is a misunderstanding of how to use view transitions, in the same way `* { will-change: transform }` is a misunderstanding. Authors should only be putting view transition names on things that need to transition. 

If we want to allow for the case where developers give everything a view transition name, I think we need a different system altogether that eliminates the need for the initial capture in cases where it isn't needed.

**Comment by noamr** [comment]
Note that the spec does allow clipping the contents of elements to the snapshot containing block (while still capturing the element and creating pseudo-elements for it), see https://www.w3.org/TR/css-view-transitions-1/#capture-rendering-characteristics-algorithm.

**Comment by khushalsagar** [comment]
^ we only clip the root snapshot to the snapshot containing block. Non-root elements are captured in entirety (modulo hardware constraints). If we were to clip element snapshots to the snapshot containing block, they'd be completely empty for offscreen elements. IIUC the examples @jakearchibald is mentioning don't want that behaviour, since the animation can bring the content onscreen we need to capture the offscreen pixels.

> I think that example is a misunderstanding of how to use view transitions, in the same way * { will-change: transform } is a misunderstanding. Authors should only be putting view transition names on things that need to transition.

That's fair but `will-change` comes with a huge disclaimer about its rendering characteristics. Since the property is explicitly about adding rendering cost for animation smoothness, authors have to realize what they're doing when they use this property. `view-transition-name`'s primary purpose is element matching. The offscreen rendering is pretty subtle so I doubt authors will think about it. In fact most will probably assume what you said above, "same as CSS animations ...". Put an animation on everything you need and the browser will deal with optimizing offscreen stuff.

Picking a default to always capture for this new property is a bit like picking `transform` as the default value for `will-change` from a perf standpoint. I realize it's more nuanced since `will-change` is purely perf, while there is visual impact based on the default we use here.

**Comment by jakearchibald** [comment]
Let's say I want to perform a view transition where I reorder a list:

- The first and last items are swapping. No other items are moving.
- The first item is north of the viewport, the last item is south of the viewport.

I would consider it a bug if the browser decides not to perform the animation I explicitly requested. It would be _really buggy_ if the browser _sometimes_ decides not to perform the animation I explicitly requested.

**Comment by khushalsagar** [comment]
> It would be really buggy if the browser sometimes decides not to perform the animation I explicitly requested.

Ok, that I see now. It would indeed be very annoying if it sometimes works *and* differently across browsers. So let's not do any UA based margin heuristic.

Let's say we pick "only capture when onscreen" as the default. Would you say it would still feel buggy if you had to add the `view-transition-inclusion-area` property in addition to `view-transition-name` if your named element can be offscreen? Because then it's part of the API contract we're setting up with this property. Just the name is not enough to say, "I want this to always animate".

**Comment by bramus** [comment]
> if you had to add the `view-transition-inclusion-area` property

Feels more like something to put as a descriptor in the `@view-transition` at-rule, as this applies to the VT as a whole instead of individual elements.

**Comment by jakearchibald** [comment]
> Let's say we pick "only capture when onscreen" as the default.

It would break all "slide in/out from the top/side/bottom" transitions.

**Comment by khushalsagar** [comment]
> Feels more like something to put as a descriptor in the @view-transition at-rule, as this applies to the VT as a whole instead of individual elements.

Wouldn't authors have cases where this is different for different elements?

> It would break all "slide in/out from the top/side/bottom" transitions.

Just to be clear, it won't break a slide in from out-of-viewport type entry (name only in new DOM) animation. If the element in the new DOM is onscreen, it'll get captured and you can make its pseudo slide in from out of viewport to its final state. Same thing for exit (name only in old DOM) animations.

But yea, the list example you gave where you swap the first and last entry, and the last entry is offscreen, would break.

> I guess advice to developers would be: Browsers are about to break your transitions. Add * { view-transition-inclusion-area: everywhere } to prevent this 😄

Ha, well if authors end up using that CSS then changing the current behaviour is moot. I feel like you're convinced that authors will know to not put a `view-transition-name` unless that element needs to be animated, i.e., it will come onscreen during the transition. And changing the default here will just get in their way.

I'm worried that won't be the case, especially as we make it easier to declaratively add names with features like https://github.com/w3c/csswg-drafts/issues/8320. But I also don't have any other idea for avoiding this perf footgun...

So let's keep the current behaviour. Also easier from an implementor standpoint, no compat risk to deal with. :)

**Comment by vmpstr** [comment]
I think it's reasonable to consider making animations that begin and end off-screen not animate by default (and hear me out).

It is indeed going to become easier to mark things as participating in a view transition, via auto naming and auto matching. There are also thoughts about making something like transition: content where a view transition may implicitly start on an element when something about it changes.

All of these features would make it easy to have animations happening completely off-screen. And by easy I mean even without the author realizing it. To Khushal's point, this is going to likely regress performance more than a CSS animation would (although I would challenge that offscreen CSS animations are "free"). This is due to the fact that view transitions include visual captures.

If by default, we don't run an animation if the start and the end states are off-screen, then in currently Chromium implementation we would still run a capture, but we can get rid of it quickly. We can also see if we can figure out some implementation optimization for this, but I can't think of anything off top the of my head. In any case, we'd at least have a possibility of optimizing this.

What we lose is the list reorder case, where a thing intentionally flies through the viewport. That's ok? We can add a property that opts into the behavior of always capturing regardless of off-screen-ness

**Comment by khushalsagar** [comment]
@vmpstr trying to tease the API contact you're proposing. The browser is free to not generate pseudo-elements for a `view-transition-name` in the old DOM if:

- The named element is offscreen in the old DOM.
- There is no corresponding name in the new DOM *or* the named element in the new DOM is also offscreen on first render.

For the case where a name is only in the new DOM, the browser can optimize out its rendering as usual since we're producing live snapshots.

Not generating pseudo-elements is the web observable difference here. So an author can't bring the pseudo-element onscreen only during the transition. Internally an engine can optimize by caching any representation of the old DOM until the new DOM is ready.

**Comment by jakearchibald** [comment]
Another case to consider is https://simple-set-demos.glitch.me/raf-dust-2/

```js
// Dummy element for the dust
for (const name of ['dust-1', 'dust-2']) {
  const div = document.createElement("div");
  div.style.viewTransitionName = name;
  div.style.contain = "paint";
  document.body.append(div);  
}
```

These elements are created to produce groups that can be used within the transition, but they don't relate to any real content. They're only in the "old" view.

It feels like this would become unreliable under the proposed behaviour.

**Comment by noamr** [comment]
A potential perhaps-legit optimization the UA can do is this:
- See that a particular URL (or session) consistently marks `view-transition-name` to elements that are never in the viewport
- Start not capturing the snapshot for these elements, opportunistically 
- If this pattern is broken, and one of those elements goes into the viewport, skip this particular transition and turn the optimization off for this URL/session. So you'd have one prematurely skipped transition for the whole session if that ever happened but you'd gain the optimization.

However, this type of optimization is a tangent to this discussion.

The thing I think we should solve here is - allowing the web developer to decide that for particular view-transition participants, if it's offscreen in one of the phases, the animation should act like an exit/entry transition. "I would prefer this article to fade in rather than to fly from the bottom if it comes from outside the viewport". The only way to achieve this today is by measuring/interscecting with javascript, and IMO it's an important enough use-case to make declarative.

**Comment by bramus** [comment]
> A potential perhaps-legit optimization the UA can do is this:
> …

This feels like a bit too much hand wavy magic. I’d rather see authors explicitly opt-in to this, like they do with `[loading=lazy]` and `content-visibility: auto;`.

**Comment by jakearchibald** [comment]
> The thing I think we should solve here is - allowing the web developer to decide that for particular view-transition participants, if it's offscreen in one of the phases, the animation should act like an exit/entry transition.

Yes, that's the feature developers are asking for, and the use-case that was identified in the previous issue to this one https://github.com/WICG/view-transitions/issues/84

**Comment by noamr** [comment]
> > A potential perhaps-legit optimization the UA can do is this:
> > …
> 
> This feels like a bit too much hand wavy magic. I’d rather see authors explicitly opt-in to this, like they do with `[loading=lazy]` and `content-visibility: auto;`.

Agreed, using it as an example of things that implementations can do without changing the spec, which doesn't solve the problem presented in this issue.

**Comment by khushalsagar** [comment]
> The thing I think we should solve here is - allowing the web developer to decide that for particular view-transition participants, if it's offscreen in one of the phases, the animation should act like an exit/entry transition.

That's fair. This is the primary use-case for this issue. That said, because this property is defining the behaviour for offscreen elements in View Transitions it intersects with the UA optimizing offscreen animations. Even if we don't change the default, we'd want authors to use this property to opt-out of caching offscreen content if they know the transition won't bring it onscreen.

> It feels like this would become unreliable under the proposed behaviour.

This property can always be used to explicitly specify whether an element is captured irrespective of its viewport relative position. So the author can opt-in to any behaviour (which would reliably happen across browsers) if the default doesn't work for them.

Just to reiterate, my concern is limited to caching a representation of old DOM elements. As per spec, the author can bring an image of a named element in the old DOM in the viewport at any point. So the UA must generate and cache it for the lifetime of the transition. And while it's visually correct, I'm not sure if authors need that behaviour in most cases or understand the additional performance cost associated with it. There is no such issue with the new DOM snapshots because they come from a live element. The UA can already decide whether to paint them based on the position of the pseudo-element rendering that image.

**Comment by noamr** [comment]
Sorry to backtrack, but I think there is a problem we haven't considered with this, unrelated to the optimization options.

Everything is fine when the viewport stays where it is, but if we implement this suggestion it might produce undesired results when scrolling (or otherwise changing anything about the viewport).

Imagine the following scenario:
- An element flies into the top of the viewport.
- We don't want this animation to occur if the element is out of the viewport, so we mark it as `offscreen: hide` or whatever.
- The user scroll down during the animation
- The parent of the element also has a `view-transition-name` (it could be the root)

There are 3 options here:
1. If we don't capture that element at all, it would potentially appear double! as part of capturing the old root, and as the new element
2. If we don't capture the element, but also ignore it in the parent, the element would disappear immediately when scrolling, and the new element would re-appear
3. we can capture both states, but add a hint that lets the author use a specific animation when flying from out-of-viewport. This way, the author can decide, for example, to fade out the old element and fade in the new one.

(2) is perhaps OK but feels less flexible. We can achieve (3) in a "soft" way, by using UA view-transition-classes, e.g.:

```css
::view-transition-group(header.-ua-from-out-of-viewport) {
  animation: fade-in-instead-of-fly;
}
```

**Comment by khushalsagar** [comment]
> The user scroll down during the animation

At what stage of the transition is this happening? For both the old and new DOM we'll compute the element's visibility only once when we're looking for elements with names. So a visibility change after that (from scrolling) won't impact whether the element is captured or the pseudo-DOM structure.

And can you clarify in what situation the UA would add "-ua-from-out-of-viewport". The idea of UA provided classes sounds reasonable, just want to make sure I get the details right.

**Comment by noamr** [comment]
> > The user scroll down during the animation
> 
> 
> 
> At what stage of the transition is this happening? For both the old and new DOM we'll compute the element's visibility only once when we're looking for elements with names. So a visibility change after that (from scrolling) won't impact whether the element is captured or the pseudo-DOM structure.

The old element is out of viewport when capturing the old state, and the usef scrolls to a position where it's in viewport. So we considered it invisible when capturing but that state changed.

> 
> 
> 
> And can you clarify in what situation the UA would add "-ua-from-out-of-viewport". The idea of UA provided classes sounds reasonable, just want to make sure I get the details right.

If at the time of capturing the old state the element is invisible, the pseudos would receive this vt class. The author can then decide how to animate it if at all.

**Comment by noamr** [comment]
> > The old element is out of viewport when capturing the old state, and the usef scrolls to a position where it's in viewport. So we considered it invisible when capturing but that state changed.
> 
> So the `view-transition-name` is ignored in the old DOM but it is applied in the new DOM (when scrolling brought that element into the viewport). There won't be a `::view-transition-old` pseudo for that name. If the name has an ancestor with a name 2 cases are possible:
> 
> * The named ancestor is the root in which case the old content is not captured at all (because root snapshots clip to the viewport).

Right, I guess animations with captured roots anyway don't work well with scroll.
But this element can be a child of a captured element other than root as well.

> * The named ancestor is not the root in which case the old content could be in this ancestor's snapshot.
> 
> There will be a `::view-transition-new` for it though. Would the `:only-child` pseudo-class suffice to detect this state?

That would be option (2). you'd scroll down and the element would disappear immediately rather than fade out. This is ok but perhaps not always the desired outcome.

**Comment by khushalsagar** [comment]
^ A couple of thoughts for the options:

- Another reason I like `view-transition-visibility` is that this feels conceptually similar [`position-visibility`](https://drafts.csswg.org/css-anchor-position-1/#position-visibility). Specify how view transition name should behave based on the visibility of the element. So would become a consistent pattern for authors to understand.
   But I would also be ok with `view-transition-capture` from the list above.

- For the values, I like `always` as the default (as-in always capture). `in-viewport` is nice to say "only capture when in viewport".

**Comment by khushalsagar** [comment]
> What about scrollers with clipped overflow where you’d want to use this?

Good point! How about `element-visible` or `visible`? Again drawing inspiration from `position-visibility`'s `anchors-visible`.

**Comment by bramus** [comment]
Not a fan of the term “visible” (it has many interpretations) but good to go for something that is consistent.

Another (inconsistently named) option that came to mind is `not-clipped` which clearly communicates how it should be interpreted.

**Comment by khushalsagar** [comment]
> not to elements inside viewprot but not visible

which case were you thinking about? If the element has `visibility: hidden`. I was assuming that visibility here is limited to whether the element has been clipped by an ancestor. And it looks like anchor positioning is doing the same thing [here](https://drafts.csswg.org/css-anchor-position-1/#position-visibility:~:text=If%20the%20element%20has%20a%20default%20anchor%20element%20but%20the%20anchor%20element%20is%20clipped%20by%20intervening%20elements).

**Comment by khushalsagar** [comment]
Hmmm ok. I would push back against occlusion but the cases you outlined seem reasonable.

**Comment by noamr** [comment]
I want to present a counter-argument to where this is going.
The new prop is a simple condition: "if the element is 100% outside of viewport during capture don't capture it".

This feels like it would work great as an optimization for the cases where the element is out of the viewport throughout the whole transition, and for art-direction when we want to avoid *very far* fly-ins. Perhaps that's sufficient and we may come to that conclusion! But a lot of cases may fall in between... e.g. avoiding a fly-in that is "quite far" but the element still intersects the viewport by a single pixel, or wanting to *change* the animation when an element flys in rather than not capture it at all.

It could be that a customizable margin property in addition would fix that, but something about the condition feels rigid, like a very specific solution to a particular problem that perhaps doesn't extend well to adjacent problems?

This is not an opposition to the proposed direction, but rather a call-out to check if this rigid condition we're proposing aligns well enough with real world use cases in such a way that would be useful enough to replace `IntserctionObserver`-based solutions in those cases.

**Comment by khushalsagar** [comment]
> avoiding a fly-in that is "quite far" but the element still intersects the viewport by a single pixel

IIUC the farthest animation distance in this case would be animating from 1 edge of the viewport to the other (for the longer axis). I was assuming that is not the case authors are trying to fix in that, the element is in the viewport throughout the animation. It's the cases where the element stays offscreen for majority of the animation that looks awkward. Because then instead of getting an animation, the element just "shows up" in the last few frames of the animation. I could very likely be missing an edge case!

The other reason I like this proposal is that it makes it trivial for authors to optimize away forced rendering of offscreen content that VT currently does. But the property has to be useful in most cases for authors to actually use it. So definitely happy to get more feedback on whether this actually addresses the target use-cases!

**Comment by bramus** [comment]
> It could be that a customizable margin property in addition would fix that

We could use insets to determine the capture area? Positive values for insets go inwards and negative ones go outwards.

Making up properties as I go, it would result in something like this:

```css
view-transition-capture-inset: auto | <percentage-length>{1,4}; /* To be set on the transition root upon which you call startViewTransition(). For `document`, set this on `:root` */
view-transition-visibility: always | not-clipped; /* To be set on individual elements */
```

An alternative to `view-transition-visibility` as suggested a few times here is to determine the “capture things or not” on the transition root itself. Something like:

```css
view-transition-capture-inset: auto | <percentage-length>{1,4};
view-transition-capture-mode: all | not-clipped;
```

The case against `view-transition-visibility` is that you need to do this on a per-element basis. OTOH is does give you fine-grained control.

The case against `view-transition-capture-mode` is that it’s _“all or nothing”_. OTOH it’s “set once and you’re good”.

_Side note: a hack would be to use scroll-driven animations to unset the `view-transition-name` when an element out of view, but that’s a hack that we shouldn’t really rely on here._

**Comment by khushalsagar** [comment]
We can add `view-transition-capture-inset` to customize the margin for determining an element's visibility on top of `view-transition-visibility` going forward.

Between `view-transition-visibility` and `view-transition-capture-mode`, you could do the latter with something like:

```css
#transition-root * {
  view-transition-visibility: not-clipped;
}
```

So seems nicer to give `view-transition-visibility` which allows per element choice..?

Also, just to clarify, @noamr's argument is not about how the author expresses whether the element should be captured (based on visibility) or not. But whether the API should be about controlling whether the element is captured. Or just exposing what the element's visibility was. For example, `::view-transition-old:in-viewport`/`::view-transition-old:out-of-viewport`, and let the author decide how they want to react for each case. One of which could be to override the UA default morph animation with an entry/exit animation.

My take is that the above is more complicated than it needs to be, especially given that this feature is about providing an easy declarative way to do something which is already possible in script. So the declarative solution should make ~most of the cases easy instead of trying to solve all possible cases. And also allow authors to optimize away offscreen rendering.

**Comment by khushalsagar** [comment]
One more thing, one good point which has been raised on this issue is that if we default to ignoring offscreen elements it would break legit use-cases. And authors would just do something like `* { view-transition-capture: all }` to get back the old behaviour.

I suspect most cases where authors add a name to a lot of elements is list items (like changing a grid layout), where likely only the element position is changing and its content is the same. So for the perf thing, maybe https://github.com/w3c/csswg-drafts/issues/9406 is a better fix. We can default to skip caching paint for offscreen content and require authors to explicitly opt-in. Since for new live elements, we can generate paint when the pseudo comes onscreen this won't be an issue unless a cross-fade was really needed. And I don't see why that would be necessary if the old content was offscreen to begin with.

**Comment by khushalsagar** [comment]
Offline discussion on this today. We have a proposal for how the pseudo-element rendering changes if the old element is offscreen at capture time.

### Morph Animation
This is the case where both old and new elements exist for the same name:
-  The geometry animation on the group pseudo is same as before.
- `::view-transition-old` renders the new live image instead of the old snapshot.
- Instead of setting up cross-fade animations, `::view-transition-old` has a 0->0 opacity animation and `::view-transition-new` has a 1->1 opacity animation. This is mostly to preserve existing behaviour where we use `animation-fill-mode: both` to set the final value for these pseudos.
- `-ua-mix-blend-mode-plus-lighter` is not added to the animation on old/new pseudos since we're not cross-fading them.

A few other options considered were:

- Don't generate the old pseudo at all. This will conflict with author expectations that this situation is limited to entry animations. Authors even use `:only-child` to specify custom entry animations for this mode.
- Generate the old pseudo but don't have any image there. The concern with this is that it might be a larger compat issue if authors have custom animations (like slide out the old and slide in the new). From a perf perspective, since the default UA animations is rendering old at 0 opacity we can also optimize away the rendering.

### Exit animation
This is the case where only the old element exists for a name. A couple of options here:
1. Generate the old pseudo but render no content in it. And switch from a fade out animation to a 0->0 opacity animation.
2. Don't generate any pseudos for this name at all, since we'll basically be doing animations with empty boxes.

Leaning towards 1) because again least compat risk.

---------
The question for how we define an element is offscreen is still open and can be discussed separately.

**Comment by noamr** [comment]
I think that in addition to the resolution we need to enable authors ways to override/customize this behavior:
- override the default, stating that a particular element should not have its contents clipped. This could perhaps also apply to the root element, except with a different user-agent default, e.g. `view-transition-content-overflow: auto | all | viewport` where the root defaults to `viewport` and everything else defaults to `auto.
- Customize the default animation behavior when we're displaying the new live snapshot instead of the old snapshot. Thinking something like a pseudo-class on the pseudo elements (`::view-transition-{group|old|new}(header):old-content-hidden` except with a better name?)

**Comment by khushalsagar** [comment]
> override the default, stating that a particular element should not have its contents clipped.

Did you mean to say, "a particular element *should* have its contents clipped". Because the default for everything except the root is no ancestor/viewport clipping. Agreed that we should have something like this but we can treat it as a separate problem? Just want to confirm that there's no reason a property like this should be done along with the behaviour change we resolved on.

> Customize the default animation behavior when we're displaying the new live snapshot instead of the old snapshot.

This makes sense, authors should be able to detect this state. How about `old-content-skipped` or `old-content-captured`? The name should make it clear that old content was available (this isn't an entry animation) but we explicitly didn't capture it.

We'd likely eventually add a property to opt-in to this behaviour even for cases where the old element is onscreen. All the grid re-layout type of transitions where elements move onscreen without changing content can be optimized then. So the name shouldn't make it sound like the old content was offscreen. This is where I wasn't sure if `old-content-hidden` is ambiguous.

**Comment by noamr** [comment]
> > override the default, stating that a particular element should not have its contents clipped.
> 
> Did you mean to say, "a particular element _should_ have its contents clipped". Because the default for everything except the root is no ancestor/viewport clipping. Agreed that we should have something like this but we can treat it as a separate problem? Just want to confirm that there's no reason a property like this should be done along with the behaviour change we resolved on.

No, sorry, I mean "a particular element should not have its content clipped, even if it is away from the viewport".

> 
> > Customize the default animation behavior when we're displaying the new live snapshot instead of the old snapshot.
> 
> This makes sense, authors should be able to detect this state. How about `old-content-skipped` or `old-content-captured`? The name should make it clear that old content was available (this isn't an entry animation) but we explicitly didn't capture it.

Sounds good!

> 
> We'd likely eventually add a property to opt-in to this behaviour even for cases where the old element is onscreen. All the grid re-layout type of transitions where elements move onscreen without changing content can be optimized then. So the name shouldn't make it sound like the old content was offscreen. This is where I wasn't sure if `old-content-hidden` is ambiguous.

+1

**Comment by noamr** [comment]
As far as the OP goes, this is achievable today using scroll-driven animations. See this demo: https://codepen.io/noamr-the-selector/pen/jOjPBXP

Basically the element in question has a `view-timeline-name` and an `animation-timeline` referencing itself, with `@keyframes` that animate the `view-transition-name` property. This makes it so that only when the element is in the viewport it has a `view-transition-name`. The offset can be controlled using `view-timeline-inset`.

It's perhaps hacky to use SDA for this but it works today and can be used for other things than view-transitions as well!

**Comment by jakearchibald** [comment]
> It's perhaps hacky to use SDA for this

It is. I'm sad that the OP has been ignored given the number of developers asking for this. I guess we can continue to solve it with JavaScript.

**Comment by noamr** [comment]
> > It's perhaps hacky to use SDA for this
> 
> It is. I'm sad that the OP has been ignored given the number of developers asking for this. I guess we can continue to solve it with JavaScript.

Given the following:

```css
@keyframes apply-view-transition-name-when-in-range {
  from { view-transition-name: var(--view-transition-name);  }
  to { view-transition-name: var(--view-transition-name); }
}

header {
  --view-transition-name: header;
  animation-name: apply-view-transition-name-when-in-range;
  animation-timeline: view();
}
```

Can you be more specific about what behavior the OP requests that's different? Is it just the aesthetic aspect of using an animation? If It looked something like this (with the exact same behavior) would it make any difference?

```css
header {
  @in-view {
     view-transition-name: header;
  }
}
```

Note that using view timelines for this has the added value of being able to control a different behavior when slightly off screen and far away from screen:

```css

```css
@keyframes apply-view-transition-name{
  from { 
   view-transition-name: var(--view-transition-name);  
   view-transition-class: near;
  }
  10% { view-transition-class:  inside; }
  90% { view-transition-class:  inside; }
  to { 
    view-transition-class: near;
    view-transition-name: var(--view-transition-name); 
  }
}

header {
  --view-transition-name: header;
  animation-name: apply-view-transition-name;
  view-timeline-inset: 50vmax;
  animation-timeline: view();
}
```
```

I think the problem with this issue is that it defines a scroll-driven behavior for view-transitions and we are trying to do this as a shortcut, while we already have a scroll-driven primitive that can do this, and the main issue with it is that it's tied to "animations"?.

**Comment by jakearchibald** [comment]
> Is it just the aesthetic aspect of using an animation?

Pretty much, yeah. I'm guessing it doesn't account for ink overflow either? Are Chrome's dev rels happy with documenting this feature as a hack with scroll timelines?

**Comment by noamr** [comment]
> > Is it just the aesthetic aspect of using an animation?
> 
> Pretty much, yeah. I'm guessing it doesn't account for ink overflow either? 

Correct, you have to set `view-timeline-inset` explicitly. But this is a known view-timeline issue, e.g. if you currently use text as your view timeline it wouldn't account for the overflowing edges.

> Are Chrome's dev rels happy with documenting this feature as a hack with scroll timelines?

We'll discuss this internally to see if people are content with it. I have it on my list to document it (and another hack to override the "don't capture contents if it's far from the screen" behavior we resolved on here) as ways to change view-transition behavior based on scroll position of participating elements.

**Comment by khushalsagar** [comment]
> I'm sad that the OP has been ignored given the number of developers asking for this.

@jakearchibald apologies if it came across that this feature request is being ignored. OP's use-case of customizing the UX when content is offscreen was explicitly discussed at the last CSS f2f, the detailed notes are [here](https://github.com/w3c/csswg-drafts/issues/8282#issuecomment-2165842379).

One of the questions raised in that discussion was if we should be providing the offscreen state of the old/new DOM elements as a pseudo-class. Something like `:offscreen and :offscreen(start | end | both)`. Then authors can choose to react to these states by specifying a custom animation. That seems more flexible than a property which says, "ignore this element's name if its offscreen".

**Comment by noamr** [comment]
Created #10581 and #10582 to account for the two issues with using `view-timeline` for this:
- Account for ink-overflow somehow (beyond the explicit `view-timeline-inset`), perhaps with `view-timeline-range`.
- using animation for something that's not really an animation (but just a range) is weird

I believe that these issues belong in scroll-driven animations and not in view transitions, as they define scroll-related behavior. For example, the same patterns can be used to change the behavior of a regular CSS transition if it's away from the viewport.

**Comment by jakearchibald** [comment]
> WebKit has raised an implementation concern regarding displaying the live image twice.

`-webkit-box-reflect` is already a thing 😄

**Comment by noamr** [comment]
> > WebKit has raised an implementation concern regarding displaying the live image twice.
> 
> `-webkit-box-reflect` is already a thing 😄

True, https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h#L654 seems to do something like cloning a layer but it's been a while since I touched CA so will leave this to @nt1m and @mattwoodrow

**Comment by nt1m** [comment]
`-webkit-box-reflect` isn't a pattern we'd like to repeat. It's intrusive and broken in many ways (e.g. video layers are not cloned for instance).

**Comment by noamr** [comment]
Closed because the optimization use-case seems resolved for now. #10581 and #10582 are still open to track the issue of improving the ergonomics of controlling how view-transitions behave based on distance from viewport.
